### PR TITLE
12. [5519] PNI ProductPage factories 

### LINKED ISSUE #5519: 12. BuyersGuide, SoftwareProduct and GeneralProduct page factories
> We need new factory code to auto-generate these page types for us. 

Criteria:
- [ ] Wagtail Factories for the new pages 
  - [ ] Factory for BuyersguidePage
  - [ ] Factories for Products (GeneralProductPage and SoftwareProductPage's, but don't generate generic ProductPage pages as those are used for internal relationship use)
  - [ ] Remove old factories that generate product model data

### PR DESCRIPTION
**Pull request by KalobTaulien**
Closes #5519 
Related PRs/issues #5489 
Review link: https://foundation-s-5519-pni-f-0rxa94.herokuapp.com/

## Criteria

- [ ] Wagtail Factories for the new pages 
  - [x] Factory for BuyersguidePage
  - [x] Factories for Products (GeneralProductPage and SoftwareProductPage's, but don't generate generic ProductPage pages as those are used for internal relationship use)
  - [ ] Remove old factories that generate product model data. **Note: Let's not remove old factories until we remove all the old PNI code - let's remove it all at once in the future in case we still need it in the near future**

The migration in this PR is the same one that's in #5518, I needed it for writing factories. And it will be renamed to 0018 when this gets merged in.

### CODE REVIEW

File: network-api/networkapi/wagtailpages/pagemodels/products.py [hunk 1]
```diff
@@ -343,7 +343,7 @@ class ProductPage(FoundationMetadataPageMixin, Page):
 
     @property
     def current_tally(self):
-        return sum(self.votes.get_votes())
+        return sum(self.votes.get_or_create_votes())
```
Comment (KalobTaulien) [inline network-api/networkapi/wagtailpages/pagemodels/products.py:346]: Code in this file is the same as what's been reviewed in #5518 but I needed it here as well. Comes with the 0017 migration but will be renamed to 0018 when this PR is approved and rebased.

File: network-api/networkapi/buyersguide/factory.py [hunk 2]
```diff
@@ -197,9 +206,120 @@ class Meta:
...
+    for i in range(50):
+        # Create 50 GeneralProductPages with Privacy Link Orderables
+        general_page = GeneralProductPageFactory.create(
+            parent=pni_homepage,
+        )
+        fake_privacy_policy = ProductPagePrivacyPolicyLinkFactory(
+            page=general_page
+        )
+        general_page.privacy_policy_links.add(fake_privacy_policy)
+        general_page.save_revision().publish()
```
Comment (Pomax) [inline network-api/networkapi/buyersguide/factory.py:311]: This looks like it's missing setting zero-or-more `related_products`.

File: network-api/networkapi/buyersguide/factory.py [hunk 3]
```diff
@@ -197,9 +206,120 @@ class Meta:
...
+        # Create 50 SoftwareProductPages with Privacy Link Orderables
+        software_page = SoftwareProductPageFactory.create(
+            parent=pni_homepage,
+        )
+        software_page.save_revision().publish()
+        fake_privacy_policy = ProductPagePrivacyPolicyLinkFactory(
+            page=software_page
+        )
+        software_page.privacy_policy_links.add(fake_privacy_policy)
+        software_page.save_revision().publish()
```
Comment (Pomax) [inline network-api/networkapi/buyersguide/factory.py:321]: (same as above)

File: network-api/networkapi/buyersguide/factory.py [hunk 4]
```diff
@@ -197,9 +206,120 @@ class Meta:
...
+    recording_alert = LazyFunction(get_extended_yes_no_value)
+    recording_alert_helptext = Faker('sentence')
+    medical_privacy_compliant = Faker('boolean')
+    medical_privacy_compliant_helptext = Faker('sentence')
+    host_controls = Faker('sentence')
+    easy_to_learn_and_use = Faker('boolean')
+    easy_to_learn_and_use_helptext = Faker('sentence')
+    first_published_at = Faker('past_datetime', start_date='-2d', tzinfo=timezone.utc)
+    last_published_at = Faker('past_datetime', start_date='-1d', tzinfo=timezone.utc)
+    slug = None
```
Comment (Pomax) [inline network-api/networkapi/buyersguide/factory.py:239]: If this is required, let's leave a code comment explaining why we need to explicitly set this to `None`.

**Review by Pomax — CHANGES_REQUESTED** [review]

File: network-api/networkapi/buyersguide/factory.py (hunk 2 above)
Comment (KalobTaulien) [reply network-api/networkapi/buyersguide/factory.py:313]: Added 3 related products per page

File: network-api/networkapi/buyersguide/factory.py (hunk 4 above)
Comment (KalobTaulien) [reply network-api/networkapi/buyersguide/factory.py:241]: Ahh yes. Remnants of a factory thing I was testing. Its not needed. Will fix in the next commit.

File: network-api/networkapi/wagtailpages/migrations/0017_product_voting.py [hunk 5]
```diff
@@ -10,7 +10,7 @@ class Migration(migrations.Migration):
 
     dependencies = [
         ('wagtailpages', '0015_auto_20201110_1957'),
-        ('wagtailpages', '0016_merge_20201117_1917.py'),
+        ('wagtailpages', '0016_merge_20201117_1917'),
```
Comment (Pomax) [inline network-api/networkapi/wagtailpages/migrations/0017_product_voting.py:13]: so... another migration landed in the mean time. Migrations are the best! >_>;;

Latest is 0018_auto_20201126_2017.py so this'll now be 0019

**Review by Pomax — CHANGES_REQUESTED** [review]
technically R+ but yay for migrations

File: network-api/networkapi/wagtailpages/migrations/0017_product_voting.py (hunk 5 above)
Comment (KalobTaulien) [reply network-api/networkapi/wagtailpages/migrations/0017_product_voting.py:13]: [https://github.com/mozilla/foundation.mozilla.org/tree/pni-snippet-conversion/network-api/networkapi/wagtailpages/migrations](0018 isn't in yet). We're still on 0017 in `pni-snippet-converstion.` 

When there's a migration bump I'll fix up the numbering. Thanks for keeping an eye on these!

**Review by Pomax — CHANGES_REQUESTED** [review]
I'm getting some flake8 errors here:
```
network-api/networkapi/wagtailpages/pagemodels/products.py:10:1: F401 'django.http.Http404' imported but unused
network-api/networkapi/wagtailpages/pagemodels/products.py:364:38: E231 missing whitespace after ':'
network-api/networkapi/wagtailpages/pagemodels/products.py:364:47: E231 missing whitespace after ','
network-api/networkapi/wagtailpages/pagemodels/products.py:539:32: F821 undefined name 'HttpResponseNotFound'
network-api/networkapi/wagtailpages/pagemodels/products.py:564:28: F821 undefined name 'HttpResponseNotFound'
```
with those fixed, I see a single error during testing:
```
ERROR: test_localised_about (network-api.networkapi.buyersguide.tests.AboutViewTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/network-api/networkapi/buyersguide/tests.py", line 456, in test_localised_about
    response = self.client.get(url)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/test/client.py", line 535, in get
    response = super().get(path, data=data, secure=secure, **extra)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/test/client.py", line 347, in get
    **extra,
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/test/client.py", line 422, in generic
    return self.request(**r)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/test/client.py", line 503, in request
    raise exc_value
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.7/contextlib.py", line 74, in inner
    return func(*args, **kwds)
  File "/app/network-api/networkapi/utility/redirects.py", line 21, in wrapper
    return func(request, *args, **kwargs)
  File "/app/network-api/networkapi/buyersguide/views.py", line 170, in render_view
    '*privacy not included') + ' - About',
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/shortcuts.py", line 36, in render
    content = loader.render_to_string(template_name, context, request, using=using)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/loader.py", line 62, in render_to_string
    return template.render(context, request)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/backends/django.py", line 61, in render
    return self.template.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 171, in render
    return self._render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/test/utils.py", line 96, in instrumented_test_render
    return self.nodelist.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 937, in render
    bit = node.render_annotated(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 904, in render_annotated
    return self.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/loader_tags.py", line 150, in render
    return compiled_parent._render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/test/utils.py", line 96, in instrumented_test_render
    return self.nodelist.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 937, in render
    bit = node.render_annotated(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 904, in render_annotated
    return self.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/loader_tags.py", line 150, in render
    return compiled_parent._render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/test/utils.py", line 96, in instrumented_test_render
    return self.nodelist.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 937, in render
    bit = node.render_annotated(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 904, in render_annotated
    return self.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/loader_tags.py", line 150, in render
    return compiled_parent._render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/test/utils.py", line 96, in instrumented_test_render
    return self.nodelist.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 937, in render
    bit = node.render_annotated(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 904, in render_annotated
    return self.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/loader_tags.py", line 62, in render
    result = block.nodelist.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 937, in render
    bit = node.render_annotated(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 904, in render_annotated
    return self.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/loader_tags.py", line 62, in render
    result = block.nodelist.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 937, in render
    bit = node.render_annotated(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 904, in render_annotated
    return self.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/loader_tags.py", line 188, in render
    return template.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 173, in render
    return self._render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/test/utils.py", line 96, in instrumented_test_render
    return self.nodelist.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 937, in render
    bit = node.render_annotated(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 904, in render_annotated
    return self.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/defaulttags.py", line 513, in render
    return self.nodelist.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 937, in render
    bit = node.render_annotated(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/base.py", line 904, in render_annotated
    return self.render(context)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/django/template/library.py", line 192, in render
    output = self.func(*resolved_args, **resolved_kwargs)
  File "/app/dockerpythonvenv/lib/python3.7/site-packages/wagtail/contrib/routable_page/templatetags/wagtailroutablepage_tags.py", line 23, in routablepageurl
    base_url = page.relative_url(site, request)
AttributeError: 'str' object has no attribute 'relative_url'
```

**Review by Pomax — APPROVED** [review]

**Pull request merged**